src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/stores/ResourceFormStore.js

```diff
@@ -87,15 +88,163 @@ export default class ResourceFormStore extends AbstractFormStore implements Form
+                    }
+                    if (!(oldChildData.type in newTypes)) {
+                        result[key][childKey] = this.calculateDifference(
+                            oldChildSchema,
+                            newChildSchema,
+                            oldChildData,
+                            newChildData,
+                            parentPath.concat([key, childKey])
+                        );
+                    }
+                }
+
+                continue;
+            }
+
+            if (!oldData[key] || newSchema[key].type !== oldSchema[key]?.type) {
+                // set originData
+                result[key] = newData[key];
+            }
+        }
+
+        return result;
+    }
+
+    removeObsoleteData(oldSchema: Schema, newSchema: Schema, data: Object, parentPath: string[] = []) {
+        if (!oldSchema || !newSchema) {
+            return;
+        }
+
+        for (const key in oldSchema) {
+            const {items: newItems, type: newType, types: newTypes} = newSchema[key] || {};
+            const {items: oldItems, type: oldType, types: oldTypes} = oldSchema[key];
+
+            if (newType === SECTION_TYPE && newItems &&
+                oldType === SECTION_TYPE && oldItems) {
+                this.removeObsoleteData(
+                    oldItems,
+                    newItems,
+                    data,
+                    parentPath
+                );
+                continue;
+            }
+
+            if (newTypes && oldTypes
+                && Object.keys(newTypes).length > 0 && Object.keys(oldTypes).length > 0
+                && data[key] && isArrayLike(data[key])
+            ) {
+                for (const childKey of data[key].keys()) {
+                    const childData = data[key][childKey];
+                    const oldChildSchema = oldTypes[childData.type]?.form;
+                    let newChildSchema = newTypes[childData.type]?.form;
+
+                    if (!newChildSchema) {
+                        const defaultType = newSchema[key]?.defaultType;
+                        if (defaultType) {
+                            newChildSchema = newTypes[defaultType].form;
+                            //set default type
+                            const path = parentPath.concat([key, childKey]).join('/') + '/type';
+                            this.change(path, defaultType);
+                        }
+                    }
+
+                    this.removeObsoleteData(
+                        oldChildSchema,
+                        newChildSchema,
+                        childData,
+                        parentPath.concat([key, childKey])
+                    );
+                }
+
+                continue;
+            }
+
+            if (!newSchema[key] || newSchema[key]['type'] !== oldSchema[key]['type']) {
+                // remove obsolete data from previous schema
+                this.remove(parentPath.concat(key).join('/'));
```

Comment:
I still think we need this in the frontend, otherwise the admin crashes in the following scenario:

We have template `A` with property `title` as `text_line` and template `B` with property `title` as `single_media_selection` .

If template `A` is selected with a value in `title` and the user switches to template `B` the admin crashes. For blocks we can avoid this with a switch to the `default-block-type` but for the other properties I don't see an easy way to replace it other than removing the property completely.